src/main/scala/Validator.scala

```diff
@@ -23,38 +32,173 @@ class Validator(var schemaUri: Option[String], sourceUri: String = "") {
+        )
+
+      // http://example.com/imports-data-as-csv?year=1010
+
+      val workingWithUserSpecifiedMetadata =
+        if (schemaUri.isDefined && possibleSchemaUri.toString == schemaUri.get)
+          true
+        else false
+      maybeCsvUri
+        .map(csvUri => {
+          if (
+            tableGroupContainsCsv(
+              tableGroup,
+              csvUri
+            ) || warningsAndErrors.errors.nonEmpty || workingWithUserSpecifiedMetadata
+          ) // the last 2 or cases are required. See tests 104 and 121 respectively
+            Right((tableGroup, warningsAndErrors))
+          else {
+            val message =
+              s"Schema file does not contain a definition for ${maybeCsvUri}"
+            Left(
+              SchemaDoesNotContainCsvError(
+                new IllegalArgumentException(message)
+              )
+            )
+          }
+
+        })
+        .getOrElse(Right((tableGroup, warningsAndErrors)))
+
+    } catch {
+      case metadataError: MetadataError =>
+        Left(GeneralCsvwLoadError(metadataError))
+      case e: java.io.FileNotFoundException => Left(CascadeToOtherFilesError(e))
+      case e: Throwable => {
+        Left(GeneralCsvwLoadError(e))
+      }
+    }
+  }
+
+  private def tableGroupContainsCsv(
+      tableGroup: TableGroup,
+      csvUri: URI
+  ): Boolean = {
+    val csvUrl = csvUri.toString
+    val tables = tableGroup.tables
+
+    val csvUrlWithoutQueryString = getUriWithoutQueryString(csvUri).toString
+
+    // todo: We need to be able to try both relative & absolute CSVURIs here.
+
+    tables.contains(csvUrl) || tables.contains(csvUrlWithoutQueryString)
+  }
+
+  private def getUriWithoutQueryString(csvUri: URI): URI = {
+    if (csvUri.getRawQuery == null)
+      csvUri
+    else {
+      val queryStringLength = csvUri.getRawQuery.length
+      val url = csvUri.toString
+      new URI(url.substring(0, url.length - (queryStringLength + 1)))
+    }
+  }
+
   def validate(): WarningsAndErrors = {
-    // When CSV is fetched from web and the associated schema is in the header, this wont work.
-    // Change this (returning empty warnings and errors when schemaUri is blank) when that feature is implemented
-    if (schemaUri.isEmpty) return WarningsAndErrors()
-    val absoluteSchemaUri = getAbsoluteSchemaUri(schemaUri.get)
-    Schema.loadMetadataAndValidate(absoluteSchemaUri) match {
-      case Right((tableGroup, warnings)) => {
-        if (sourceUri.isEmpty) {
-          val warningsAndErrors = validateSchemaTables(tableGroup)
-          WarningsAndErrors(
-            warningsAndErrors.warnings ++ warnings,
-            warningsAndErrors.errors
+    val absoluteSchemaUri = schemaUri.map(getAbsoluteSchemaUri)
+
+    val maybeCsvUri = csvUri.map(new URI(_))
+
+    val schemaUrisToCheck = maybeCsvUri
+      .map(csvUri =>
+        Array(
+          absoluteSchemaUri,
+          Some(new URI(s"${getUriWithoutQueryString(csvUri)}-metadata.json")),
```

Comment:
Would be good to have a link here in a comment to the relevant part of the CSV-W documentation to make it clear why we're doing this.